### PR TITLE
Make OIDC plugin public

### PR DESCRIPTION
**Pull request by korycins**
This PR makes our OIDC plugin public. It is a port of changes from private repo. 
Docs for the plugin can be found here: https://docs.saleor.io/docs/3.x/developer/available-plugins/openid-connect

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/plugins/base_plugin.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def __str__(self):
     #  Authenticate user which should be assigned to the request.
     #
     #  Overwrite this method if the plugin handles authentication flow.
-    authenticate_user: Callable[[WSGIRequest], Union["User", NoneType]]
+    authenticate_user: Callable[[WSGIRequest, Any], Union["User", NoneType]]
```
Comment (patrys) [inline saleor/plugins/base_plugin.py:140]: Any?

Comment (korycins) [reply saleor/plugins/base_plugin.py:140]: All definitions of methods in BasePlugin have `previous_value` declared as `Any`. I added it here also as it was missing previously.

Comment (patrys) [reply saleor/plugins/base_plugin.py:140]: I don't think `Any` is the correct type in any of those cases? Isn't previous value always the same type as returned value?

**Review by Qasem-h — APPROVED** [review]

File: saleor/plugins/openid_connect/utils.py [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+    if not keys:
+        logger.warning("List of JWKS keys is empty")
+    return keys
+
+
+def get_jwks_keys_from_cache_or_fetch(jwks_url: str) -> dict:
+    jwks_keys = cache.get(JWKS_KEY)
+    if jwks_keys is None:
+        jwks_keys = fetch_jwks(jwks_url)
+        cache.set(JWKS_KEY, jwks_keys, JWKS_CACHE_TIME)
```
Comment (L3str4nge) [inline saleor/plugins/openid_connect/utils.py:96]: I don't think it's good place to set keys into cache. `fetch_jwsk(url, cache=True)` might be better. I think getter method should not setting anything.

**Review by patrys — APPROVED** [review]

**Review by cmiacz — APPROVED** [review]

**Review by L3str4nge — APPROVED** [review]

File: saleor/plugins/openid_connect/plugin.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+from authlib.common.errors import AuthlibBaseError
+from authlib.integrations.requests_client import OAuth2Session
+from django.core import signing
+from django.core.exceptions import ValidationError
+from django.core.handlers.wsgi import WSGIRequest
+from jwt import ExpiredSignatureError, InvalidTokenError
+from requests import HTTPError, PreparedRequest
+
+from saleor.account.models import User
```
Comment (maarcingebala) [inline saleor/plugins/openid_connect/plugin.py:12]: 🚔  Use relative imports. 🚔

File: saleor/plugins/openid_connect/utils.py [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+            user, refresh_token, csrf_token, owner
+        )
+        tokens["csrf_token"] = csrf_token
+    return tokens
+
+
+def validate_refresh_token(refresh_token, data):
+    csrf_token = data.get("csrfToken")
+    if not refresh_token:
+        raise ValidationError(
```
Comment (maarcingebala) [inline saleor/plugins/openid_connect/utils.py:451]: IMO this shouldn't return an ValidationError but some other class, but I guess we don't have time for this now.

File: saleor/plugins/openid_connect/plugin.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+
+        user = get_or_create_user_from_payload(
+            parsed_id_token, self.config.authorization_url
+        )
+
+        user_permissions = []
+        if self.config.use_scope_permissions:
+            scope = token_data.get("scope")
+            user_permissions = self._use_scope_permissions(user, scope)
+            if not user.is_staff and bool("saleor:staff" in scope or user_permissions):
```
Comment (maarcingebala) [inline saleor/plugins/openid_connect/plugin.py:275]: `saleor:staff` could be moved to some constant. It's being used in a number of places and it's a hardcoded string everywhere.

**Comment by manojLondhe** [comment]
In which Saleor release this will be out?

**Comment by patrys** [comment]
@manojLondhe It should be part of the upcoming 3.2.

**Comment by toastedtoast** [comment]
3.2 is not mentioned in the roadmap yet. [https://github.com/orgs/saleor/projects/3/views/1](https://github.com/orgs/saleor/projects/3/views/1)
Any estimate when the first release for 3.2 may happen?

**Pull request merged**